charms/tensorboard-controller/tests/integration/test_charm.py

```diff
@@ -0,0 +1,200 @@
+    replicas = tensorboard.get("status", {}).get("readyReplicas")
+
+    resource_class_kind = resource_class.__name__
+    if replicas == 1:
+        logger.info(f"{resource_class_kind}/{resource_name} readyReplicas == {replicas}")
+    else:
+        logger.info(
+            f"{resource_class_kind}/{resource_name} readyReplicas == {replicas} (waiting for '1')"
+        )
+
+    assert replicas == 1, f"Waited too long for {resource_class_kind}/{resource_name}!"
+
+
+async def test_create_tensorboard(ops_test: OpsTest):
+    """Test Tensorboard creation."""
+    lightkube_client = Client()
+    load_in_cluster_generic_resources(lightkube_client)
+
+    # Create PVC for Tensorboard logs and Tensorboard
+    resources = codecs.load_all_yaml(
+        "\n---\n".join([PVC_TEMPLATE_FILE.read_text(), TENSORBOARD_TEMPLATE_FILE.read_text()]),
+        context={"pvc_name": PVC_NAME, "tensorboard_name": TENSORBOARD_NAME},
+    )
+    for rsc in resources:
+        lightkube_client.create(rsc, namespace=ops_test.model_name)
+
+    try:
+        tensorboard_ready = lightkube_client.get(
+            TENSORBOARD_RESOURCE,
+            name=TENSORBOARD_NAME,
+            namespace=ops_test.model_name,
+        )
+    except ApiError:
+        assert False
+    assert tensorboard_ready, f"Tensorboard {ops_test.model_name}/{TENSORBOARD_NAME} not found!"
+
+    assert_replicas(lightkube_client, TENSORBOARD_RESOURCE, TENSORBOARD_NAME, ops_test.model_name)
+
+
+@pytest.mark.abort_on_fail
+async def test_remove_with_resources_present(ops_test: OpsTest):
+    """Test remove with all resources deployed.
+
+    Verify that all deployed resources that need to be removed are removed.
+    """
+    # remove deployed charm and verify that it is deleted
+    await ops_test.model.remove_application(app_name=APP_NAME, block_until_done=True)
+    assert APP_NAME not in ops_test.model.applications
+
+    # verify that all resources that were deployed are deleted
+    lightkube_client = Client()
+
+    # verify that all created CRDs and Services in namespace are deleted
+    for rsc in [
+        CustomResourceDefinition,
+        Service,
+    ]:
+        rsc_list = lightkube_client.list(
+            rsc,
+            labels=[("app.juju.is/created-by", APP_NAME)],
+            namespace=ops_test.model_name,
+        )
+        assert not list(rsc_list)
+
+    # verify that all created Pods in namespace are deleted
+    for rsc in [
+        Pod,
+    ]:
+        rsc_list = lightkube_client.list(
+            rsc,
+            labels=[("app", TENSORBOARD_NAME)],
+            namespace=ops_test.model_name,
+        )
+        assert not list(rsc_list)
+
+    # verify that the PVC and Tensorboard are deleted
+    try:
+        for rsc, name in [
+            (PersistentVolumeClaim, PVC_NAME),
+            (TENSORBOARD_RESOURCE, TENSORBOARD_NAME),
+        ]:
+            _ = lightkube_client.get(rsc, name=name, namespace=ops_test.model_name)
```

Comment:
I think this test is missing an assertion.  If the objects exist, I think this passes and the test succeeds
```suggestion
            _ = lightkube_client.get(rsc, name=name, namespace=ops_test.model_name)
            assert False, f"Found leftover resource {name} of type {rsc}"
```

Reply:
Added your suggestion! I also added a manual deletion of the PVC created for the purposes of this test, since it won't be removed by Juju (doesn't know about it) or K8s (no owner refs, so it doesn't care about it).